File: utils/validation.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Security: Input validation and sanitization utilities
 */

// Email validation regex
const EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

// URL validation regex
const URL_REGEX = /^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)$/;

// Social media handle validation
const HANDLE_REGEX = /^@[a-zA-Z0-9_.]{1,30}$/;

/**
 * Validates email format
 */
export const validateEmail = (email: string): { isValid: boolean; error?: string } => {
  if (!email || email.trim().length === 0) {
    return { isValid: false, error: 'Email is required' };
  }
  
  if (!EMAIL_REGEX.test(email)) {
    return { isValid: false, error: 'Invalid email format' };
  }
  
  if (email.length > 254) {
    return { isValid: false, error: 'Email is too long' };
  }
  
  return { isValid: true };
};

/**
 * Validates URL format
 */
export const validateUrl = (url: string): { isValid: boolean; error?: string } => {
  if (!url || url.trim().length === 0) {
    return { isValid: false, error: 'URL is required' };
  }
  
  if (!URL_REGEX.test(url)) {
    return { isValid: false, error: 'Invalid URL format' };
  }
  
  if (url.length > 2048) {
    return { isValid: false, error: 'URL is too long' };
  }
  
  return { isValid: true };
};

/**
 * Validates social media handle
 */
export const validateHandle = (handle: string): { isValid: boolean; error?: string } => {
  if (!handle || handle.trim().length === 0) {
    return { isValid: false, error: 'Handle is required' };
  }
  
  if (!HANDLE_REGEX.test(handle)) {
    return { isValid: false, error: 'Invalid handle format (use @username)' };
  }
  
  return { isValid: true };
};

/**
 * Validates password strength
 */
export const validatePassword = (password: string): { isValid: boolean; error?: string } => {
  if (!password || password.length === 0) {
    return { isValid: false, error: 'Password is required' };
  }
  
  if (password.length < 8) {
    return { isValid: false, error: 'Password must be at least 8 characters' };
  }
  
  if (!/(?=.*[a-z])/.test(password)) {
    return { isValid: false, error: 'Password must contain at least one lowercase letter' };
  }
  
  if (!/(?=.*[A-Z])/.test(password)) {
    return { isValid: false, error: 'Password must contain at least one uppercase letter' };
  }
  
  if (!/(?=.*\d)/.test(password)) {
    return { isValid: false, error: 'Password must contain at least one number' };
  }
  
  if (!/(?=.*[@$!%*?&])/.test(password)) {
    return { isValid: false, error: 'Password must contain at least one special character' };
  }
  
  return { isValid: true };
};

/**
 * Sanitizes text input to prevent XSS
 */
export const sanitizeText = (text: string): string => {
  if (!text) return '';
  
  return text
    .replace(/[<>]/g, '') // Remove potential HTML tags
    .replace(/javascript:/gi, '') // Remove javascript: protocol
    .replace(/on\w+=/gi, '') // Remove event handlers
    .trim();
};

/**
 * Sanitizes and validates search input
 */
export const sanitizeSearchInput = (input: string): string => {
  if (!input) return '';
  
  return input
    .replace(/[<>]/g, '')
    .replace(/javascript:/gi, '')
    .replace(/on\w+=/gi, '')
    .replace(/['"]/g, '') // Remove quotes
    .trim()
    .substring(0, 100); // Limit length
};

/**
 * Validates company name
 */
export const validateCompanyName = (name: string): { isValid: boolean; error?: string } => {
  if (!name || name.trim().length === 0) {
    return { isValid: false, error: 'Company name is required' };
  }
  
  if (name.length > 100) {
    return { isValid: false, error: 'Company name is too long' };
  }
  
  if (!/^[a-zA-Z0-9\s&.,'-]+$/.test(name)) {
    return { isValid: false, error: 'Company name contains invalid characters' };
  }
  
  return { isValid: true };
};

/**
 * Generic input validator
 */
export const validateInput = (
  value: string, 
  type: 'email' | 'url' | 'handle' | 'password' | 'text' | 'company',
  required: boolean = true,
  minLength: number = 0,
  maxLength: number = 1000
): { isValid: boolean; error?: string } => {
  
  if (required && (!value || value.trim().length === 0)) {
    return { isValid: false, error: 'This field is required' };
  }
  
  if (value && value.length < minLength) {
    return { isValid: false, error: `Minimum length is ${minLength} characters` };
  }
  
  if (value && value.length > maxLength) {
    return { isValid: false, error: `Maximum length is ${maxLength} characters` };
  }
  
  switch (type) {
    case 'email':
      return validateEmail(value || '');
    case 'url':
      return validateUrl(value || '');
    case 'handle':
      return validateHandle(value || '');
    case 'password':
      return validatePassword(value || '');
    case 'company':
      return validateCompanyName(value || '');
    default:
      return { isValid: true };
  }
};

File: utils/validation.py
"""
@license
SPDX-License-Identifier: Apache-2.0
"""

"""
Validation utilities for Python Flask application
"""

import re
from typing import Dict, Any

def validate_email(email: str) -> Dict[str, Any]:
    """
    Validate email format
    
    Args:
        email: Email address to validate
        
    Returns:
        Dictionary with validation result
    """
    if not email:
        return {
            'is_valid': False,
            'error': 'Email is required'
        }
    
    # Basic email regex
    email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    
    if not re.match(email_pattern, email):
        return {
            'is_valid': False,
            'error': 'Invalid email format'
        }
    
    if len(email) > 254:
        return {
            'is_valid': False,
            'error': 'Email is too long'
        }
    
    return {
        'is_valid': True,
        'error': None
    }

def validate_input(value: str, input_type: str, required: bool = True, 
                 min_length: int = 0, max_length: int = 1000) -> Dict[str, Any]:
    """
    Generic input validation
    
    Args:
        value: Value to validate
        input_type: Type of input
        required: Whether input is required
        min_length: Minimum length
        max_length: Maximum length
        
    Returns:
        Dictionary with validation result
    """
    if required and not value:
        return {
            'is_valid': False,
            'error': f'{input_type} is required'
        }
    
    if value and len(value) < min_length:
        return {
            'is_valid': False,
            'error': f'{input_type} must be at least {min_length} characters'
        }
    
    if value and len(value) > max_length:
        return {
            'is_valid': False,
            'error': f'{input_type} must be less than {max_length} characters'
        }
    
    return {
        'is_valid': True,
        'error': None
    }

def sanitize_text(text: str) -> str:
    """
    Sanitize text input to prevent XSS and injection attacks
    
    Args:
        text: Text to sanitize
        
    Returns:
        Sanitized text
    """
    if not text:
        return ''
    
    # Remove HTML tags
    text = re.sub(r'<[^>]+>', '', text)
    
    # Remove JavaScript protocol
    text = re.sub(r'javascript:', '', text, flags=re.IGNORECASE)
    
    # Remove event handlers
    text = re.sub(r'on\w+\s*=', '', text, flags=re.IGNORECASE)
    
    # Remove quotes
    text = text.replace('"', '').replace("'", '')
    
    # Remove CSS expressions
    text = re.sub(r'expression\s*\(', '', text, flags=re.IGNORECASE)
    
    # Strip whitespace
    text = text.strip()
    
    return text